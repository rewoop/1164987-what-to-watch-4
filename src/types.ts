export interface Film {
  id: number;
  filmTitle: string;
  filmVideo: string;
  filmGenre: string;
  releaseDate: number;
  backgroundPoster: string;
  filmPoster: string;
  ratingScore: number;
  ratingLevel: number;
  ratingCount: string;
  filmDescription: string;
  filmDirector: string;
  filmStarring: string[];
  filmRunTime: string;
  isFavoriteFilm: boolean;
}

export type Films = Film[];

export interface FilmComment {
  id: number;
  user: {
    id: number;
    name: string;
  };
  comment: string;
  date: string;
  rating: number;
}

export type FilmComments = FilmComment[];

export interface PostReview {
  rating: number;
  comment: string;
}
