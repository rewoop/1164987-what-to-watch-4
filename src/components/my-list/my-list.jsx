import React, {PureComponent} from "react";
import PropTypes from "prop-types";
import {Link} from "react-router-dom";
import {AppRoute} from "../../const.js";
import Footer from "../footer/footer.jsx";
import FilmCard from "../film-card/film-card.jsx";
import withVideo from "../../hocs/with-video/with-video";
import {Operation as DataOperation} from "../../reducer/data/data.js";
import {getFavoriteFilms} from "../../reducer/data/selectors.js";
import {connect} from "react-redux";

const FilmCardWrapped = withVideo(FilmCard);

class MyList extends PureComponent {
  componentDidMount() {
    this.props.loadFavoriteFilms();
  }

  render() {
    const {films} = this.props;

    return (
      <div className="user-page">
        <header className="page-header user-page__head">
          <div className="logo">
            <Link to={AppRoute.ROOT} className="logo__link">
              <span className="logo__letter logo__letter--1">W</span>
              <span className="logo__letter logo__letter--2">T</span>
              <span className="logo__letter logo__letter--3">W</span>
            </Link>
          </div>

          <h1 className="page-title user-page__title">My list</h1>


          <div className="user-block">
            <Link to={AppRoute.MY_LIST}>
              <div className="user-block__avatar">
                <img src="img/avatar.jpg" alt="User avatar" width="63" height="63"/>
              </div>
            </Link>
          </div>

        </header>

        <section className="catalog">
          <h2 className="catalog__title visually-hidden">Catalog</h2>

          <div className="catalog__movies-list">
            {films.length > 0 ?
              films.map((currentFilm) => {
                return <FilmCardWrapped
                  key={currentFilm.filmTitle}
                  film={currentFilm}
                />;
              }) :
              <h1 style={{marginLeft: `250px`}}>Add movies to favorites to watch later</h1>}
          </div>
        </section>

        <Footer />
      </div>
    );
  }

}

MyList.propTypes = {
  films: PropTypes.arrayOf(
      PropTypes.shape({
        filmTitle: PropTypes.string,
        filmVideo: PropTypes.string,
        filmGenre: PropTypes.string,
        releaseDate: PropTypes.number,
        backgroundPoster: PropTypes.string,
        filmPoster: PropTypes.string,
        ratingScore: PropTypes.number,
        ratingLevel: PropTypes.number,
        ratingCount: PropTypes.string,
        filmDescription: PropTypes.string,
        filmDirector: PropTypes.string,
        filmStarring: PropTypes.arrayOf(
            PropTypes.string
        ),
        runTime: PropTypes.string,
      })
  ).isRequired,
  loadFavoriteFilms: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => {
  return {
    films: getFavoriteFilms(state),
  };
};

const mapDispatchToProps = (dispatch) => ({
  loadFavoriteFilms() {
    dispatch(DataOperation.loadFavoriteFilms());
  }
});

export {MyList};
export default connect(mapStateToProps, mapDispatchToProps)(MyList);
