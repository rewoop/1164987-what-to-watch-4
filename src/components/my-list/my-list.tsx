import * as React from "react";
import {Link} from "react-router-dom";
import {AppRoute} from "../../const";
import Footer from "../footer/footer";
import FilmCard from "../film-card/film-card";
import withVideo from "../../hocs/with-video/with-video";
import {Operation as DataOperation} from "../../reducer/data/data";
import {getFavoriteFilms} from "../../reducer/data/selectors";
import {connect} from "react-redux";

interface Props {
  films: {
    id: number,
    filmTitle: string,
    filmVideo: string,
    filmGenre: string,
    releaseDate: number,
    backgroundPoster: string,
    filmPoster: string,
    ratingScore: number,
    ratingLevel: number,
    ratingCount: string,
    filmDescription: string,
    filmDirector: string,
    filmStarring: string[],
    runTime: string,
    isFavoriteFilm: boolean
  }[],
  loadFavoriteFilms: () => void,
}

const FilmCardWrapped = withVideo(FilmCard);

class MyList extends React.PureComponent<Props> {
  componentDidMount() {
    this.props.loadFavoriteFilms();
  }

  render() {
    const {films} = this.props;

    return (
      <div className="user-page">
        <header className="page-header user-page__head">
          <div className="logo">
            <Link to={AppRoute.ROOT} className="logo__link">
              <span className="logo__letter logo__letter--1">W</span>
              <span className="logo__letter logo__letter--2">T</span>
              <span className="logo__letter logo__letter--3">W</span>
            </Link>
          </div>

          <h1 className="page-title user-page__title">My list</h1>


          <div className="user-block">
            <Link to={AppRoute.MY_LIST}>
              <div className="user-block__avatar">
                <img src="img/avatar.jpg" alt="User avatar" width="63" height="63"/>
              </div>
            </Link>
          </div>

        </header>

        <section className="catalog">
          <h2 className="catalog__title visually-hidden">Catalog</h2>

          <div className="catalog__movies-list">
            {films.length > 0 ?
              films.map((currentFilm) => {
                return <FilmCardWrapped
                  key={currentFilm.filmTitle}
                  film={currentFilm}
                />;
              }) :
              <h1 style={{marginLeft: `250px`}}>Add movies to favorites to watch later</h1>}
          </div>
        </section>

        <Footer />
      </div>
    );
  }

}

const mapStateToProps = (state) => {
  return {
    films: getFavoriteFilms(state),
  };
};

const mapDispatchToProps = (dispatch) => ({
  loadFavoriteFilms() {
    dispatch(DataOperation.loadFavoriteFilms());
  }
});

export {MyList};
export default connect(mapStateToProps, mapDispatchToProps)(MyList);
