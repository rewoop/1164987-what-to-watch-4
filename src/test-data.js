export const TestFilm = {
  id: 1,
  filmTitle: `The Rock`,
  filmGenre: `Thriller`,
  filmVideo: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  releaseDate: 1996,
  isFavoriteFilm: true,
  runTime: `356`,
  backgroundPoster: `img/bg-the-grand-budapest-hotel.jpg`,
  filmPoster: `img/the-grand-budapest-hotel-poster.jpg`,
  ratingScore: 9,
  ratingLevel: 9,
  ratingCount: `1337 ratings`,
  filmDescription: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege.
  Gustave prides himself on providing first-class service to the hotel's guests, including satisfying the sexual needs of the many elderly women who stay there. When one of Gustave's lovers dies mysteriously, Gustave finds himself the recipient of a priceless painting and the chief suspect in her murder.`,
  filmDirector: `Wes Andreson`,
  filmStarring: [
    `Bill Murray`,
    `Edward Norton`,
    `Jude Law`,
    `Willem Dafoe`,
    `Tom Cruz`
  ],
};

export const TestFilms = [TestFilm];

export const TestComment = {
  id: 0,
  user: {
    id: 1,
    name: `Kate Muir`,
  },
  comment: `Discerning travellers and Wes Anderson fans will luxuriate in the glorious Mittel-European kitsch of one of the director's funniest and most exquisitely designed movies in years.`,
  date: `December 24, 2016`,
  rating: 9,
};

export const TestComments = [TestComment];
