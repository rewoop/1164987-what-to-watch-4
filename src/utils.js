import {ALL_GENRES, RatingLevel} from "./const.js";

export const formatReviewDate = (dateString) => {
  const date = new Date(dateString);
  const dateTimeFormat = new Intl.DateTimeFormat(`en`, {year: `numeric`, month: `2-digit`, day: `2-digit`});
  const [{value: month},, {value: day},, {value: year}] = dateTimeFormat.formatToParts(date);
  return `${year}-${month}-${day}`;
};

export const extend = (a, b) => {
  return Object.assign({}, a, b);
};

export const getGenresList = (films) => {
  return [ALL_GENRES].concat(Array.from(new Set(films.map((film) => film.filmGenre))));
};

export const formatRatingScoreToLevel = (score) => {
  if (score >= RatingLevel.MIN_BAD && score <= RatingLevel.MAX_BAD) {
    return `Bad`;
  } else if (score > RatingLevel.MAX_BAD && score <= RatingLevel.MAX_NORMAL) {
    return `Normal`;
  } else if (score > RatingLevel.MAX_NORMAL && score <= RatingLevel.MAX_GOOD) {
    return `Good`;
  } else if (score > RatingLevel.MAX_GOOD && score < RatingLevel.MAX_VERY_GOOD) {
    return `Very good`;
  } else {
    return `Awesome`;
  }
};
