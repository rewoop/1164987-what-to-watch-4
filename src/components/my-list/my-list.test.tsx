import * as React from "react";
import * as renderer from "react-test-renderer";
import {MyList} from "./my-list";
import {Provider} from "react-redux";
import configureStore from "redux-mock-store";
import NameSpace from "../../reducer/name-space";
import thunk from 'redux-thunk';
import {noop} from "../../utils";
import {Films} from "../../types";
import {TestFilms} from "../../test-data";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

jest.mock(`react-router-dom`);

const filmsInfo: Films = TestFilms;

it(`Should My List render correctly`, () => {
  const store = mockStore({
    [NameSpace.DATA]: {
      favoriteFilms: filmsInfo
    }
  });

  const tree = renderer.create(
      <Provider store={store}>
        <MyList
          films={filmsInfo}
          loadFavoriteFilms={noop}
        />
      </Provider>
  );
  expect(tree.toJSON()).toMatchSnapshot();
});
