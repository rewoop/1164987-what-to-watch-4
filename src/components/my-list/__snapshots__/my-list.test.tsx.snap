// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should My List render correctly 1`] = `
<div
  className="user-page"
>
  <header
    className="page-header user-page__head"
  >
    <div
      className="logo"
    />
    <h1
      className="page-title user-page__title"
    >
      My list
    </h1>
    <div
      className="user-block"
    />
  </header>
  <section
    className="catalog"
  >
    <h2
      className="catalog__title visually-hidden"
    >
      Catalog
    </h2>
    <div
      className="catalog__movies-list"
    >
      <article
        className="small-movie-card catalog__movies-card"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <h3
          className="small-movie-card__title"
        />
      </article>
    </div>
  </section>
  <footer
    className="page-footer"
  >
    <div
      className="logo"
    />
    <div
      className="copyright"
    >
      <p>
        Â© 2019 What to watch Ltd.
      </p>
    </div>
  </footer>
</div>
`;
